File: ts/indicators/john-ehlers/instantaneous-trend-line/length-params.ts
import { BarComponent } from '../../../entities/bar-component';
import { QuoteComponent } from '../../../entities/quote-component';
import { TradeComponent } from '../../../entities/trade-component';

/** Describes parameters to create an instance of the InstantaneousTrendLine indicator based on length. */
export interface InstantaneousTrendLineLengthParams {
  /**
   * Length is the length (the number of time periods, \u2113) of the Instantaneous Trend Line.
   *
   * The smoothing factor \u03b1 is derived as \u03b1 = 2/(\u2113+1).
   * The value should be a positive integer, greater or equal to 1.
   */
  length: number;

  /**
   * A component of a bar to use when updating the indicator with a bar sample.
   *
   * If _undefined_, the bar component defaults to BarComponent.Median (median price)
   * and will be shown in the indicator mnemonic as 'hl/2' since it differs from the
   * framework default (close).
   */
  barComponent?: BarComponent;

  /**
   * A component of a quote to use when updating the indicator with a quote sample.
   *
   * If _undefined_, the quote component will have a default value and will not be shown in the indicator mnemonic.
   */
  quoteComponent?: QuoteComponent;

  /**
   * A component of a trade to use when updating the indicator with a trade sample.
   *
   * If _undefined_, the trade component will have a default value and will not be shown in the indicator mnemonic.
   */
  tradeComponent?: TradeComponent;
}

export function defaultLengthParams(): InstantaneousTrendLineLengthParams {
    return { length: 28 };
}

File: go/indicators/johnehlers/instantaneoustrendline/lengthparams.go
package instantaneoustrendline

import "zpano/entities"

// LengthParams describes parameters to create an instance of the Instantaneous Trend Line indicator
// based on length.
type LengthParams struct {
	// Length is the length, ℓ, of the instantaneous trend line.
	// The equivalent smoothing factor α is:
	//
	//	α = 2/(ℓ + 1), 0 < α ≤ 1, 1 ≤ ℓ
	//
	// The value should be a positive integer, greater or equal to 1.
	Length int

	// BarComponent indicates the component of a bar to use when updating the indicator with a bar sample.
	//
	// If zero, the default (BarMedianPrice) is used and the component is not shown in the indicator mnemonic.
	BarComponent entities.BarComponent

	// QuoteComponent indicates the component of a quote to use when updating the indicator with a quote sample.
	//
	// If zero, the default (QuoteMidPrice) is used and the component is not shown in the indicator mnemonic.
	QuoteComponent entities.QuoteComponent

	// TradeComponent indicates the component of a trade to use when updating the indicator with a trade sample.
	//
	// If zero, the default (TradePrice) is used and the component is not shown in the indicator mnemonic.
	TradeComponent entities.TradeComponent
}

// DefaultLengthParams returns a [LengthParams] value populated with Ehlers defaults.
func DefaultLengthParams() *LengthParams {
	return &LengthParams{
		Length: 28,
	}
}
